File: validate_ids.js
const fs = require('fs');

async function run() {
    console.log("Starting strict ID validation using Fuzzwork API...");
    
    // Read current data
    const fileContent = fs.readFileSync('src/data/pi_data.js', 'utf8');
    const commoditiesMatch = fileContent.match(/export const commodities = (\[[\s\S]*?\]);/);
    if (!commoditiesMatch) {
        console.error("Could not find commodities array");
        return;
    }
    
    let commodities;
    try {
        commodities = eval(commoditiesMatch[1]);
    } catch(e) {
        console.error("Failed to parse commodities");
        return;
    }

    // Collect all names to verify
    const names = commodities.map(c => c.name);
    
    // Fuzzwork allows multiple typenames separated by |
    // But let's be safe and do batches of 20
    const nameToNewId = {};
    const oldIdToName = {};

    for (let i = 0; i < names.length; i += 20) {
        const batch = names.slice(i, i + 20);
        const queryStr = batch.map(n => encodeURIComponent(n)).join('|');
        const url = `https://www.fuzzwork.co.uk/api/typeid.php?typename=${queryStr}`;
        
        try {
            const res = await fetch(url);
            const data = await res.json();
            
            for (const item of data) {
                if (item.typeID && item.typeID !== 0) {
                    nameToNewId[item.typename] = item.typeID;
                } else {
                    console.log(`WARNING: Fuzzwork returned 0 or null for ${item.typename}`);
                }
            }
        } catch (e) {
            console.error("Fetch failed for batch", batch);
        }
    }

    let mismatchCount = 0;

    // Apply new IDs
    for (const item of commodities) {
        oldIdToName[item.id] = item.name;
        
        const correctId = nameToNewId[item.name];
        if (correctId) {
            if (item.id !== correctId) {
                console.log(`FIXED: ${item.name} (${item.id} -> ${correctId})`);
                item.id = correctId;
                mismatchCount++;
            }
        } else {
            console.log(`ERROR: Could not resolve ID for ${item.name}`);
        }
    }

    // Update inputs using the name mapping to ensure references don't break
    for (const item of commodities) {
        if (item.inputs) {
            for (const input of item.inputs) {
                const inputName = oldIdToName[input.id];
                if (inputName && nameToNewId[inputName]) {
                    const correctInputId = nameToNewId[inputName];
                    if (input.id !== correctInputId) {
                        input.id = correctInputId;
                        mismatchCount++;
                    }
                }
            }
        }
    }

    if (mismatchCount === 0) {
        console.log("SUCCESS: All commodities already have the correct official SDE Type IDs!");
        return;
    }

    // Reconstruct the file string
    let newCommoditiesString = "[\n";
    let currentTier = "";

    for (const item of commodities) {
        if (item.tier !== currentTier) {
            currentTier = item.tier;
            newCommoditiesString += `\n    // --- ${currentTier} Commodities ---\n`;
        }
        const inputsStr = item.inputs.map(i => `{ id: ${i.id}, quantity: ${i.quantity} }`).join(', ');
        newCommoditiesString += `    { id: ${item.id}, name: "${item.name}", tier: "${item.tier}", inputs: [${inputsStr}], outputYield: ${item.outputYield} },\n`;
    }
    newCommoditiesString += "];";

    const beforeStr = fileContent.substring(0, commoditiesMatch.index);
    const afterStr = fileContent.substring(commoditiesMatch.index + commoditiesMatch[0].length);
    
    fs.writeFileSync('src/data/pi_data.js', beforeStr + `export const commodities = ${newCommoditiesString}` + afterStr);
    
    console.log(`Done! Fixed ${mismatchCount} total ID mismatches in pi_data.js!`);
}

run();

File: validate_ids.py
import json
import urllib.request
import urllib.parse
import re

def run():
    print("Starting validation using Python...")
    with open('src/data/pi_data.js', 'r', encoding='utf-8') as f:
        content = f.read()
        
    # Extract commodities array block
    match = re.search(r'export const commodities = (\[.*?\]);', content, re.DOTALL)
    if not match:
        print("Could not find commodities array")
        return
        
    array_str = match.group(1)
    
    # We will use a quick regex to extract all names
    names = re.findall(r'name:\s*"([^"]+)"', array_str)
    # Deduplicate
    names = list(set(names))
    
    name_to_id = {}
    
    # Fetch in batches of 10
    for i in range(0, len(names), 10):
        batch = names[i:i+10]
        query = "|".join([urllib.parse.quote(n) for n in batch])
        url = f"https://www.fuzzwork.co.uk/api/typeid.php?typename={query}"
        try:
            req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
            with urllib.request.urlopen(req) as response:
                data = json.loads(response.read().decode())
                for item in data:
                    if item.get('typeID', 0) != 0:
                        name_to_id[item['typename']] = item['typeID']
        except Exception as e:
            print(f"Failed to fetch batch {batch}: {e}")
            
    print(f"Fetched {len(name_to_id)} correct IDs.")
    
    mismatch_count = 0
    new_content = content
    
    # Now we need to carefully replace IDs.
    # The safest way is to parse the file line by line and update the 'id: xxx' 
    # but ONLY when we know which name it belongs to.
    
    lines = content.split('\n')
    for idx, line in enumerate(lines):
        if 'name: "' in line and 'tier:' in line:
            # It's a commodity definition line
            name_match = re.search(r'name:\s*"([^"]+)"', line)
            if name_match:
                name = name_match.group(1)
                correct_id = name_to_id.get(name)
                if correct_id:
                    # Find current ID
                    current_id_match = re.search(r'\{\s*id:\s*(\d+)', line)
                    if current_id_match:
                        current_id = int(current_id_match.group(1))
                        if current_id != correct_id:
                            print(f"FIXED: {name} ({current_id} -> {correct_id})")
                            lines[idx] = re.sub(r'\{\s*id:\s*\d+', f'{{ id: {correct_id}', line)
                            mismatch_count += 1

    # What about input IDs?
    # We must replace input IDs based on the old->new mapping
    # But we don't have the old mapping for everything easily without evaluating.
    # Since we only care about images loading, fixing the main definition IDs is step 1.
    
if __name__ == "__main__":
    run()
